web: Refactor radio buttons -> RadioButton

unclejustin opened the pull request:
Refactor files in `client/web/src/insights` to use Wildcard `RadioButton` component.

Related to https://github.com/sourcegraph/sourcegraph/issues/24235

unclejustin commented on client/web/src/insights/components/form/form-radio-input/FormRadioInput.tsx:
```diff
@@ -1,7 +1,11 @@
+
 interface RadioInputProps extends InputHTMLAttributes<HTMLInputElement> {
+    /** Name of radio input. */
+    name: string
```
Resetting name here to make it required.

unclejustin commented on client/web/src/insights/components/form/form-radio-input/FormRadioInput.tsx:
```diff
@@ -15,18 +19,32 @@ interface RadioInputProps extends InputHTMLAttributes<HTMLInputElement> {
+    const radioProps = { name, value, checked, onChange, disabled }
     return (
-        <label
+        <span
```
`label` is  now included in `RadioButton`

vovakulikov replied:
Hmm. That's a bummer. Cause as I understand radio button title and description are not clickable anymore. I think we should put title and description content as children of the `RadioButton` component to make it clickable

unclejustin replied:
Fixed. I had to take some more "creative license" with `RadioButton`, but Ihink it's for the best really. See notes in the `RadioButton` component.

unclejustin commented on client/wildcard/src/components/Form/RadioButton/RadioButton.tsx:
```diff
@@ -19,6 +22,22 @@ export type RadioButtonProps = {
-export const RadioButton: React.FunctionComponent<RadioButtonProps> = React.forwardRef((props, reference) => (
-    <BaseControlInput {...props} type="radio" ref={reference} />
-))
+export const RadioButton: React.FunctionComponent<RadioButtonProps> = React.forwardRef(
+    ({ isValid, labelProps, children, ...props }, reference) => {
+        const isValidStyle = getValidStyle(isValid)
+
+        return (
+            <label {...labelProps} className={classNames('form-check-label', isValidStyle, labelProps.className)}>
+                <input
+                    {...props}
+                    ref={reference}
+                    type="radio"
+                    className={classNames('form-check-input', isValidStyle, props.className)}
+                />
+                {children}
+            </label>
+        )
+    }
+)
```
`BaseControlInput` has `label` as a sibling to the input, which is not how we use radio buttons currently. Blissfully nobody is using `RadioButton` yet (that I could find). Tagging @sourcegraph/frontend-platform to double check this as I may be wildly out of line 😁

vovakulikov reviewed (approved):
LGTM from the code insights side. Left one comment about the label thing.

unclejustin commented:
@vovakulikov asking you to take a look at this again. I refactored it quite a bit to make it much simpler.

Specifically, I've removed `label` from `RadioButton` completely. There are enough edge cases and layout possibilities that it does not make sense to keep them together IMO. This has the add on benefit fo addressing your previous concern.

vovakulikov commented on client/wildcard/src/components/Form/RadioButton/RadioButton.tsx:
```diff
@@ -19,6 +21,20 @@ export type RadioButtonProps = {
+    ({ name, isValid, ...props }, reference) => {
+        const isValidStyle = getValidStyle(isValid)
+
+        return (
```
It's truly simple. But It actually seems super simple to be in the separate component then. 

as I see our abstraction for form radio input seems this
```tsx
<label
    data-placement={labelTooltipPosition}
    data-tooltip={labelTooltipText}
     className={classnames('d-flex flex-wrap align-items-center', className, {
        'text-muted': disabled,
    })}
  >
    <RadioButton {...radioProps} name={name} aria-label={title} />
    <span className="pl-2">{title}</span>

   {description && (
       <>
         <span className="pl-2 pr-2">—</span>
         <span className="text-muted">{description}</span>
      </>
    )}
 </label>
 ```
 
In this case, we have to put some custom value as label content and I remember that the first version of this PR radion button had a label internally via BaseControlInput component which was blocking us from simple reuse right? 

But what if we change the `RadionButton` component in a way with label wraps input like we have in the code insights base 

```tsx
const { checked, value, aria-label, name, /* other input specific props */, ...otherProps }
<label {... otherProps}>
  <input  type="radio" {checked, value, aria-label, name} />
  {children}
</label>
```

I think this would be better rather than we have at the moment in the `BaseCotnrolInput` component. 
Because 
- We don't require input id to connect label and input
- We allow any children with consumer custom logic in case if we need to render some content with conditions.

Our code insights radio button would be something like this

```tsx
<RadioButton
    data-placement={labelTooltipPosition}
    data-tooltip={labelTooltipText}
     className={classnames('d-flex flex-wrap align-items-center', className, {
        'text-muted': disabled,
    })}
    {... radioProps}
    aria-label={title
  >
    <span className="pl-2">{title}</span>

   {description && (
       <>
         <span className="pl-2 pr-2">—</span>
         <span className="text-muted">{description}</span>
      </>
    )}
 </RadioButton >
 ```

vovakulikov reviewed (commented):
It's truly simple. But It actually seems super simple to be in the separate component then. 

as I see our abstraction for form radio input seems this
```tsx
<label
    data-placement={labelTooltipPosition}
    data-tooltip={labelTooltipText}
     className={classnames('d-flex flex-wrap align-items-center', className, {
        'text-muted': disabled,
    })}
  >
    <RadioButton {...radioProps} name={name} aria-label={title} />
    <span className="pl-2">{title}</span>

   {description && (
       <>
         <span className="pl-2 pr-2">—</span>
         <span className="text-muted">{description}</span>
      </>
    )}
 </label>
 ```
 
In this case, we have to put some custom value as label content and I remember that the first version of this PR radion button had a label internally via BaseControlInput component which was blocking us from simple reuse right? 

But what if we change the `RadionButton` component in a way with label wraps input like we have in the code insights base 

```tsx
const { checked, value, aria-label, name, /* other input specific props */, ...otherProps }
<label {... otherProps}>
  <input  type="radio" {checked, value, aria-label, name} />
  {children}
</label>
```

I think this would be better rather than we have at the moment in the `BaseCotnrolInput` component. 
Because 
- We don't require input id to connect label and input
- We allow any children with consumer custom logic in case if we need to render some content with conditions.

Our code insights radio button would be something like this

```tsx
<RadioButton
    data-placement={labelTooltipPosition}
    data-tooltip={labelTooltipText}
     className={classnames('d-flex flex-wrap align-items-center', className, {
        'text-muted': disabled,
    })}
    {... radioProps}
    aria-label={title
  >
    <span className="pl-2">{title}</span>

   {description && (
       <>
         <span className="pl-2 pr-2">—</span>
         <span className="text-muted">{description}</span>
      </>
    )}
 </RadioButton >
 ```

@unclejustin It's a good chance that I missed something there. Proof me wrong if it's the case 😸.

@sourcegraph/frontend-platform please take a look at this since we want to change the Radio button component in the wildcard package.

unclejustin commented:
> But what if we change the RadionButton component in a way with label wraps input like we have in the code insights base

@vovakulikov after going to sleep I woke up with this same thought 😂  So yeah, let's do that 💪

unclejustin commented on client/wildcard/src/components/Form/internal/BaseControlInput.tsx:
```diff
@@ -10,26 +10,49 @@ export const BASE_CONTROL_TYPES = ['radio', 'checkbox'] as const
 
 export type ControlInputProps = AccessibleFieldProps<React.InputHTMLAttributes<HTMLInputElement>> &
     React.RefAttributes<HTMLInputElement> & {
+        /**
+         * The id used to match the label to the input.
+         */
+        id: string
+        /**
+         * The label for this input.
+         */
+        label: React.ReactNode
```
I've made id and label required here. We need to have all inputs labeled correctly. To do that we need an id and label text. [This article](https://gomakethings.com/revisting-aria-label-versus-a-visually-hidden-class/) explains why actual labels are better than `aria-label` better than I could.

This has the add on benefit of simplifying the render logic just a big.

vovakulikov replied:
I don't have any strong preference about label id or label and input as a child but just out of curiosity, do we have some disadvantages with label -> input structure. 

I thought that this layout doesn't need any id attribute to be accessible and still valid HTML 
```html
<label> 
  <input type='radio' />
  <span>Any content actually </span>
</label>
```

umpox replied:
Yeah so, initial context: A lot of the extra logic and type safety here are to enforce that either:

`aria-label`
or 
`label` and `id` (for `htmlFor`) 

is provided.

umpox replied:
That article makes some interesting points, but it makes me a little uneasy to be displaying visually hidden `span` tags to support screen readers. It seems the issues that the author raises are that `aria-label` isn't always announced consistently, or translated - which appears to be an issue with certain screen readers themselves rather than something we should aim to work around. 

From the little experience I have with screen readers, they have incredibly powerful functionality - and I'd be unsure if certain functions are still supported with this new approach (e.g. is it possible to list all options in a radio group through a single command? Does that still work with visually hidden labels?).

`aria-label` still seems like the best option to me, [recommended by WCAG](https://www.w3.org/TR/WCAG20-TECHS/ARIA14.html) and more likely to be future-proof. Also, @limitedmage recently [ran into some issues](https://sourcegraph.slack.com/archives/C0190QFK49X/p1629223896006000) using visually-hidden elements instead of aria-label

umpox replied:
@vovakulikov Yep, you don't need an `id` when writing HTML that way - the only issue IIRC is that Bootstrap has some frustrating sibling CSS that relies on the label being alongside the input in order to be styled correctly. When we eventually remove Bootstrap we'll have more freedom

vovakulikov replied:
Oh boy. Bootstrap does something bad again.

unclejustin commented on client/wildcard/src/components/Form/internal/BaseControlInput.tsx:
```diff
@@ -10,26 +10,49 @@ export const BASE_CONTROL_TYPES = ['radio', 'checkbox'] as const
          * The <input> type. Use one of the currently supported types.
          */
         type?: typeof BASE_CONTROL_TYPES[number]
+        /**
+         * Used to pass props through to the input element.
+         */
+        inputProps?: React.InputHTMLAttributes<HTMLInputElement>
+        /**
+         * Used to pass props through to the label element.
+         */
+        labelProps?: React.LabelHTMLAttributes<HTMLLabelElement>
```
We need to be able to customize the input and label directly. These are escape hatches to let us pass props through directly.

unclejustin commented on client/wildcard/src/components/Form/internal/BaseControlInput.tsx:
```diff
@@ -10,26 +10,45 @@ export const BASE_CONTROL_TYPES = ['radio', 'checkbox'] as const
                 ref={reference}
                 type={type}
-                className={classNames('form-check-input', getValidStyle(isValid), className)}
-                {...props}
+                className={classNames('form-check-input', getValidStyle(isValid), inputProps?.className)}
+                {...inputProps}
             />
-            {'label' in props && (
-                <FormFieldLabel htmlFor={props.id} className="form-check-label">
-                    {props.label}
-                </FormFieldLabel>
-            )}
+            <FormFieldLabel
+                {...labelProps}
+                htmlFor={id}
+                className={classNames('form-check-label', labelProps?.className)}
+            >
+                {label}
+            </FormFieldLabel>
```
Most of this is just passing through props and classes appropriately. I've also removed the check for `label` and added in id for the input.

unclejustin commented on client/wildcard/src/components/Form/internal/FormFieldLabel.tsx:
```diff
@@ -11,8 +11,13 @@ export interface FormFieldLabelProps {
  */
-export const FormFieldLabel: React.FunctionComponent<FormFieldLabelProps> = ({ htmlFor, className, children }) => (
-    <label htmlFor={htmlFor} className={className}>
+export const FormFieldLabel: React.FunctionComponent<FormFieldLabelProps> = ({
+    htmlFor,
+    className,
+    children,
+    ...props
+}) => (
+    <label htmlFor={htmlFor} className={className} {...props}>
```
Passing props through.

Side note, I don't understand why this component even exists. It is literally just a straight passthrough for `label`. The only difference is that `htmlFor` is required here, but not on `label`. I don't personally think that is worth creating another customer component for.

valerybugakov replied:
Interesting question. Was `htmlFor` the main reason for creating this component, @umpox?

umpox replied:
Yes, this component isn't vital, it's just an internal helper to enforce extra type-safety here as our current Bootstrap styles require the label to be a sibling for these elements, so `htmlFor` is required

valerybugakov replied:
Maybe it makes sense to add a comment about Bootstrap constraint here for clarification 👍

unclejustin commented:
@vovakulikov hey look I'm _still_ working on this PR. I feel like this is a final version at this point though.

Tagging @valerybugakov directly for review because I changed `BaseControlInput` quite a bit. Please let me know if I'm out of line on any of these changes 🙇

umpox commented on client/web/src/enterprise/insights/pages/insights/creation/search-insight/components/form-color-input/FormColorInput.tsx:
```diff
@@ -37,25 +39,23 @@ export const FormColorInput: React.FunctionComponent<FormColorInputProps> = memo
+                            checked: value === colorInfo.color,
+                            onChange,
+                        }}
+                        labelProps={{
+                            style: { color: colorInfo.color },
+                            title: colorInfo.name,
+                        }}
+                        label={<span className={styles.formColorPickerRadioControl} />}
```
Trying to understand this further

Do we need `title`? I've previously seen `title` used as a fallback when `label` is unable to be used for whatever reason. Seems strange to see `title` on a `label` element.

Is this accessible? The empty span won't add any content to screen readers. I feel like it'd be better here to use `colorInfo.name` within `label`. Then, we could also just add `labelClassName` [here](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/wildcard/src/components/Form/internal/AccessibleFieldType.ts#L19:3) instead of spreading `labelProps`!

valerybugakov replied:
@umpox's suggestion sounds reasonable. Are we missing something important about `title`?

umpox reviewed (commented):
Left some comments, thank for raising these discussions!

valerybugakov commented on client/wildcard/src/components/Form/internal/BaseControlInput.tsx:
```diff
@@ -10,26 +10,45 @@ export const BASE_CONTROL_TYPES = ['radio', 'checkbox'] as const
+        /**
+         * Used to pass props through to the input element.
+         */
+        inputProps?: React.InputHTMLAttributes<HTMLInputElement>
```
If it's possible to [remove](https://github.com/sourcegraph/sourcegraph/pull/24709/files#r706071128) `labelProps` in favor of `labelClassname` then it should be possible to remove  `inputProps` and get the from the top level: 

```ts
const { id, label, type, className, labelClassname, inputClassname, ...inputProps } = prop
```

What is the downside of this approach?

Otherwise, should we move `AccessibleFieldProps<React.InputHTMLAttributes<HTMLInputElement>>` from the top level of `ControlInputProps` to `inputProps`?

valerybugakov reviewed (commented):
Thanks for digging into this! Forms, especially these base controls, are always hard to get right for every use case.

eseliger commented:
As part of #wg-delete-week I am going to close this PR. If it’s still needed, feel free to reopen and restore the branch.
